byzcoin/tx_pipeline.go

```diff
@@ -0,0 +1,394 @@
+
+	// try to create a new state
+	newState := func() *txProcessorState {
+		if err != nil {
+			return &txProcessorState{
+				inState.sst,
+				inState.scs,
+				append(inState.txs, TxResult{tx, false}),
+			}
+		}
+		return &txProcessorState{
+			sstOut,
+			append(inState.scs, scsOut...),
+			append(inState.txs, TxResult{tx, true}),
+		}
+	}()
+
+	// we're within the block size, so return one state
+	if s.GetBlockSize() > newState.size() {
+		return []*txProcessorState{newState}, nil
+	}
+
+	// if the new state is too big, we split it
+	newStates := []*txProcessorState{inState.copy()}
+	if err != nil {
+		newStates = append(newStates, &txProcessorState{
+			inState.sst,
+			inState.scs,
+			[]TxResult{{tx, false}},
+		})
+	} else {
+		newStates = append(newStates, &txProcessorState{
+			sstOut,
+			scsOut,
+			[]TxResult{{tx, true}},
+		})
+	}
+	return newStates, nil
+}
+
+// ProposeBlock basically calls s.createNewBlock which might block. There is
+// nothing we can do about it other than waiting for the timeout.
+func (s *defaultTxProcessor) ProposeBlock(state *txProcessorState) error {
+	config, err := LoadConfigFromTrie(state.sst)
+	if err != nil {
+		return err
+	}
+	_, err = s.createNewBlock(s.scID, &config.Roster, state.txs)
+	return err
+}
+
+func (s *defaultTxProcessor) GetInterval() time.Duration {
+	bcConfig, err := s.LoadConfig(s.scID)
+	if err != nil {
+		log.Error(s.ServerIdentity(), "couldn't get configuration - this is bad and probably "+
+			"a problem with the database! "+err.Error())
+		return defaultInterval
+	}
+	return bcConfig.BlockInterval
+}
+
+func (s *defaultTxProcessor) GetLatestGoodState() *txProcessorState {
+	st, err := s.getStateTrie(s.scID)
+	if err != nil {
+		// A good state must exist because we're working on a known
+		// skipchain. If there is an error, then the database must've
+		// failed, so there is nothing we can do to recover so we
+		// panic.
+		panic(fmt.Sprintf("failed to get a good state: %v", err))
+	}
+	return &txProcessorState{
+		sst: st.MakeStagingStateTrie(),
+	}
+}
+
+func (s *defaultTxProcessor) GetBlockSize() int {
+	bcConfig, err := s.LoadConfig(s.scID)
+	if err != nil {
+		log.Error(s.ServerIdentity(), "couldn't get configuration - this is bad and probably "+
+			"a problem with the database! "+err.Error())
+		return defaultMaxBlockSize
+	}
+	return bcConfig.MaxBlockSize
+}
+
+func (s *defaultTxProcessor) Stop() {
+	close(s.stopCollect)
+}
+
+type txPipeline struct {
+	wg        sync.WaitGroup
+	processor txProcessor
+}
+
+func (p *txPipeline) start(initialState *txProcessorState, stopChan chan bool) {
+	ctxChan, stopChan1 := p.collectTx()
+	stopChan2 := p.processTxs(ctxChan, initialState)
+
+	// wait for signal and then stop
+	<-stopChan
+	close(stopChan1)
+	close(stopChan2)
+	p.processor.Stop()
+	p.wg.Wait()
+}
+
+func (p *txPipeline) collectTx() (<-chan ClientTransaction, chan<- bool) {
+	stopChan := make(chan bool, 1)
+	outChan := make(chan ClientTransaction, 200)
```

Comment:
what's the reasoning behind 200 ?

Reply:
I'm thinking about a crazy scenario where you get tons of TXs (more than 200) which means that `outChan` will block at some point (a lot of small TXs but a slow contract) and then you stop everything. I might be wrong but I don't think that close can hang so `processTxs` will stop and `collectTx` will be stuck in line 287.

Reply:
200 is kinda arbitrary, as we do more testing hopefully we'll settle down on a suitable value. yes, `collectTx` could get stuck. I'm not too sure how we should handle the situation when clients flood the ledger. Perhaps we set a large channel size, then do a non-blocking write with a select such that if clients try to overflow the transaction queue, then we just drop the transactions.

Reply:
Ok so let's resume:
1. CollectTx is sending TX one by one in the outChan inside a goroutine
2. ProcessTx is getting the TX one by one coming from outChan inside a goroutine

What if `processTx` doesn't rely on a stop channel but only on the outChan so that when this channel closes, it stops the processing go routine because you know for sure none are coming anymore ?

Reply:
yes, that could be a simplification. then we don't need the `stopChan` for `processTxs` anymore